cql3: convert the SELECT clause evaluation phase to expressions

avikivity opened the pull request:
SELECT clause components (selectors) are currently evaluated during query execution
using a stateful class hierarchy. This state is needed to hold intermediate state while
aggregating over multiple rows. Because the selectors are stateful, we must re-create
them each query using a selector_factory hierarchy.

We'd like to convert all of this to the unified expression evaluation machinery, so we can
have just one grammar for expressions, and just one way to evaluate expressions, but
the statefulness makes this complex.

In commit 59ab9aac44f "(Merge 'functions: reframe aggregate functions in terms
of scalar functions' from Avi Kivity)", we made aggregate functions stateless, moving
their state to aggregate_function_selector::_accumulator, and therefore into the
class hierarchy we're addressing now. Another reason for keeping state is that selectors
that aren't aggregated capture the first value they see in a GROUP BY group.

Since expressions can't contain state directly, we break apart expressions that contain
aggregate functions into two: an inner expression that processes incoming rows within
a group, and an outer expression that generates the group's output. The two expressions
communicate via a newly introduced expression element: a temporary.

The problem of non-aggregated columns requiring state is solved by encapsulating
those columns in an internal aggregate function, called the "first" function.

In terms of performance, this series has little effect, since the common case of selectors
that only contain direct column references without transformations is evaluated via a fast
path (`simple_selection`). This fast-path is preserved with almost no changes.

While the series makes it possible to start to extend the grammar and unify expression
syntaxes, it does not do so. The grammar is unchanged. There is just one breaking change:
the `SELECT JSON` statement generates json object field names based on the input selectors.
In one case the name of the field has changed, but it is an esoteric case (where a function call
is selected as part of `SELECT JSON`), and the new behavior is compatible with Cassandra.

avikivity commented:
Note: the first patch is unrelated but needed to build on my machine.

@gusev-p the new `temporary` thing can be used to replace lwt_cache_id.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2183/

avikivity commented:
v2:
 - removed foreign first commit
 - removed empty commit (rebase artifact)
 - removed use of C++ syntax not available on clang 15 (constructing a struct as my_struct(x, y))

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2185/

avikivity commented:
v3:
 - adjusted #include so headers are self-contained

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2188/

nyh commented on cql3/selection/selection.cc:
```diff
@@ -297,19 +297,38 @@ class selection_with_processing : public selection {
+                        ret.push_back(agg.aggregation_function);
+                    }
+                    if (agg.state_to_result_function) {
+                        ret.push_back(agg.state_to_result_function);
```
What about the state_reduction_function - don't you need it too?

avikivity replied:
No, it's never used when evaluating a query. This is also what selectors_with_processing::used_functions() returns.

The reductions are never prepared, always materialized at the point of evaluation. There's a case to do it differently, but I don't want to mix that into the series.

nyh reviewed (changes requested):
Looks mostly good, I have some minor comments and requests.
This is a HUGE patch set, so while I did spend significant time on reading it, I'm not sure I would have found every problem. I hope our test coverage in this area is good enough.

nyh commented on cql3/expr/expression.hh:
```diff
@@ -406,13 +411,18 @@ struct usertype_constructor {
     friend bool operator==(const usertype_constructor&, const usertype_constructor&) = default;
 };
 
+struct temporary {
```
After reading the next patches, I understand better what these "temporaries" are and why their name makes sense (it's a value that can be changed, not just read). So maybe the name is good, just a comment would be helpful.

avikivity replied:
Added comments.

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -463,6 +464,45 @@ aggregate_fcts::make_min_function(data_type io_type) {
+    io_type = io_type->without_reversed().shared_from_this();
+    // The function's state is a one-element tuple containing the value, if the tuple
+    // itself is null then the an input hasn't been seen yet
+    auto state_type = data_type(tuple_type_impl::get_instance({io_type}));
```
Yes. We need to distinguish between "we saw a first value, and it was NULL" and "we did not see a first value". The comment was supposed to explain it, but maybe it's lacking.

nyh replied:
Ok, I did see the comment but wasn't sure if you used a tuple just because of this reason, or there is another purpose for this tuple - perhaps to keep multiple values. I guess there isn't.

I still think the more "proper" way to do this first() aggregator was to have an accumulator being a tuple (key, value), where the null key (not null tuple) means nothing yet, and the reducer also works. But I guess it's not important enough to make this first() perfect.

avikivity replied:
It's possible to do that indeed. But I don't think it's worth the effort, it will practically never happen.

avikivity commented on cql3/functions/first_function.hh:
```diff
@@ -0,0 +1,18 @@
+// Copyright (C) 2023-present ScyllaDB
+// SPDX-License-Identifier: (AGPL-3.0-or-later and Apache-2.0)
```
Yes

avikivity replied:
Fixed

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -2759,6 +2761,142 @@ verify_no_aggregate_functions(const expression& expr, std::string_view context_f
+            return max_over_range(c.children);
+        },
+        [] (const binary_operator& bo) {
+            return std::max(aggregation_depth(bo.lhs), aggregation_depth(bo.rhs));
```
Good idea. But I think outside the scope of this series. Note that practically all expression functions are recursive.

nyh replied:
I'll open an issue about it.

avikivity commented on cql3/expr/expr-utils.hh:
```diff
@@ -352,5 +352,18 @@ bool has_only_eq_binops(const expression&);
 
-
-}
+// How deep aggregations are nested. e.g. sum(avg(count(col))) == 3
```
The function is oblivious to CQL limitations. It just counts the nesting of aggregate functions.

I was easier for me to think of the abstract problem without adding full knowledge of the environment.

avikivity commented on cql3/expr/expr-utils.hh:
```diff
@@ -352,5 +352,18 @@ bool has_only_eq_binops(const expression&);
+//
+// becomes
+//
+//    my_agg(sum(first(x)), first(first(y)))
```
Here depth is the input. Yes, aggregation_depth(the_original_expresion) == 2. But this new function is independent of the aggregation depth of its input.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -2759,6 +2761,142 @@ verify_no_aggregate_functions(const expression& expr, std::string_view context_f
+                    .func = functions::aggregate_fcts::make_first_function(type_of(ret)),
+                    .args = {std::move(ret)},
+                });
+                desired_depth -= 1;
```
"desired_depth" has a separate copy for each recursive invocation. So this adjustment doesn't leak into other cases.

I'll convert the examples into a few unit tests. I intended to, but got exhausted by the series.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -2759,6 +2761,142 @@ verify_no_aggregate_functions(const expression& expr, std::string_view context_f
+                on_internal_error(expr_logger, "expression aggregation depth exceeds expectations");
+            }
+            desired_depth -= this_function_depth;
+            recurse_over_range(fc.args);
```
Once we exit the recursion, we also exit levellize_aggregation_depth() and the local parameter is destroyed. The capture-by-reference only holds for a single invocation.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -1033,7 +1033,8 @@ std::ostream& operator<<(std::ostream& os, const expression::printer& pr) {
             .expr_to_print = expr,
-            .debug_mode = pr.debug_mode
+            .debug_mode = pr.debug_mode,
+            .for_metadata = pr.for_metadata,
```
How about if I rename it to column_metadata (and add comments)? I agree "metadata"is too generic, but nobody will read the comments.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -1080,7 +1081,11 @@ std::ostream& operator<<(std::ostream& os, const expression::printer& pr) {
+                if (pr.for_metadata) {
+                    fmt::print(os, "{}", col.col->name_as_text());
+                } else {
+                    fmt::print(os, "{}", cql3::util::maybe_quote(col.col->name_as_text()));
```
Changed.

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -1728,10 +1730,27 @@ std::unique_ptr<prepared_statement> select_statement::prepare(data_dictionary::d
     // using `forward_service`.
     auto can_be_forwarded = [&] {
-        return selection->is_aggregate()        // Aggregation only
+        return all_aggregates(prepared_selectors)   // Note: before we levellized aggregation depth
```
It's not allowed by PostgresQL (and I assume not allowed by SQL). I assume it was introduced by mistake and then codified.

It's documented in ./docs/cql/dml.rst:

"If a column is selected without an aggregate function, in a statement with a GROUP BY, the first value encounter[sic] in each group will be returned."

Note that SQL does allow naming one of the group's keys, but that doesn't need to be aggregated.

```
# select x, y, sum(z) from t3 group by x;
ERROR:  column "t3.y" must appear in the GROUP BY clause or be used in an aggregate function
LINE 1: select x, y, sum(z) from t3 group by x;

# select x, sum(z) from t3 group by x;
 x | sum 
---+-----
(0 rows)
```

kostja replied:
It is allowed in MySQL and a few other vendors, and I believe was enabled in the latest ANSI.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -1130,7 +1130,10 @@ std::ostream& operator<<(std::ostream& os, const expression::printer& pr) {
                         },
                         [&] (const shared_ptr<functions::function>& fn) {
-                            if (!pr.for_metadata) {
+                            if (!pr.debug_mode && fn->name() == cql3::functions::aggregate_fcts::first_function_name()) {
```
first_function is artificial, so I want to present it only in debug mode.

user mode or metadata mode should omitit.

avikivity commented:
- fixed license on first_function.hh
- renamed/commented :metadata mode
- use column_definition::name_as_cql_string() instead of longer expression
- added comments about the role of expr::temporary
- improve printing of prepared function call expressions in user mode (to assist unit test) (new patch)
- add unit test for aggregation depth helpers
- fixed laconic changelog on one patch

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2207/

avikivity commented:
Flaky test, filed https://github.com/scylladb/scylladb/issues/14495, restarted CI.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2209/

avikivity commented:
Something this large should have collected more reviews.

nyh commented:
> Something this large should have collected more reviews.

Because of its largeness (I spent almost half a day reviewing it...), I wasn't optimistic that any more reviewers will jump on it. Also, nothing prevents a reviewer from reviewing it now, and asking for followup or the whole thing to be reverted.

My thinking in committing this PR "quickly" (4 days...) was that:
1. In such massive changes, good test coverage is much more important than reviews, especially when the code was already not a disaster (as an understatement), and if it has problems, they would be small and not easy to easy to spot.
2. You and @cvybhu continuously improve things in this area of the code, so the longer we "sit" on these massive refactoring PRs that don't get in, the harder it becomes to work in parallel.

cvybhu commented:
I will try to review it today. On Monday and Tuesday I focused on materialized views, but now I'm preempting that and moving back to CQL.
I think giving a week of time for people to review would be nice, there's always a lot of things to do.

cvybhu reviewed (commented):
Alright I looked through it, just enough to have a general idea of what's going on.

Looks good to me, I was positively surprised that it takes so little code to analyze and implement aggregations :)

avikivity commented on cql3/selection/selection.cc:
```diff
@@ -224,7 +225,11 @@ class selection_with_processing : public selection {
     }
 
     virtual bool is_count() const override {
-        return _factories->does_count();
+        return _selectors.size() == 1
+            && expr::find_in_expression<expr::function_call>(_selectors[0], [] (const expr::function_call& fc) {
+                auto& func = std::get<shared_ptr<cql3::functions::function>>(fc.func);
+                return func->name() == functions::function_name::native_function(functions::aggregate_fcts::COUNT_ROWS_FUNCTION_NAME);
+            });
```
Right, probably should use expr::as_if
